controllers/cluster_controller.go

```diff
@@ -64,21 +64,42 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 }
 
 func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *capiv1alpha3.Cluster, log logr.Logger) (ctrl.Result, error) {
+	// ignore cluster from being deleted if ignore annotation is set
 	if _, ok := cluster.Annotations[ignoreClusterDeletion]; ok {
 		IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
 		log.Info(fmt.Sprintf("Found annotation %s. Cluster %s/%s will be ignored for deletion", ignoreClusterDeletion, cluster.Namespace, cluster.Name))
 		return ctrl.Result{}, nil
+	}
 
+	// check if cluster has a keep-until label with a valid ISO date string
+	var keepUntilExpired bool
+	if v, ok := cluster.Labels[keepUntil]; ok {
+		t, err := time.Parse(keepUntilTimeLayout, v)
+		if err != nil {
+			ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+			log.Error(err, fmt.Sprintf("failed to parse keep-until label value for cluster %s/%s", cluster.Namespace, cluster.Name))
+			return ctrl.Result{}, nil
+		}
+		if time.Now().UTC().After(t) {
+			keepUntilExpired = true
+		} else {
+			IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+			log.Info(fmt.Sprintf("Found label %s. Cluster %s/%s will be ignored for deletion", keepUntil, cluster.Namespace, cluster.Name))
+			return ctrl.Result{RequeueAfter: 24 * time.Hour}, nil
+		}
+	} else {
+		keepUntilExpired = true
 	}
```

Comment:
```suggestion
	// check if cluster has a keep-until label with a valid ISO date string
	if v, ok := cluster.Labels[keepUntil]; ok {
		t, err := time.Parse(keepUntilTimeLayout, v)
		if err != nil {
			ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
			log.Error(err, fmt.Sprintf("failed to parse keep-until label value for cluster %s/%s", cluster.Namespace, cluster.Name))
			return ctrl.Result{}, nil
		}
		if time.Now().UTC().Before(t) {
			IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
			log.Info(fmt.Sprintf("Found label %s. Cluster %s/%s will be ignored for deletion", keepUntil, cluster.Namespace, cluster.Name))
			return ctrl.Result{RequeueAfter: 24 * time.Hour}, nil
		}
	} 
```

I'm not 100% sure if this logic is correct but I think it results in the same outcome as we're returning if we've not expired, right?
(would need to change line 102 also)